### PR TITLE
Adagrad optimizer

### PR DESCRIPTION
**Pull request by eugene-kharitonov**
* Doesn't support sparse tensors (just as e.g. Adam impl)
* Generalized XOR UT to accept other optimizers/models so that can test Adagrad and SGD in a shared code
(not sure what would be the best UT, feel free to suggest another way to test)

### CODE REVIEW

File: autograd.cpp [hunk 1]
```diff
@@ -386,6 +386,39 @@ void SGD::init_state() {
...
+/// https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py
+void Adagrad::step() {
+  for (auto& pair : model_->parameters()) {
+    auto& name = pair.first;
+    auto& grad = pair.second.grad();
+    auto& p = pair.second.data();
+    if (!grad.defined()) continue;
+
+    auto d_p = grad.data();
+    d_p.add_(p, weight_decay_);
```
Comment (ebetica) [inline autograd.cpp]: Wrap this in an if statement to not launch a CUDA kernel if weight_decay_ is 0.

File: test.cpp [hunk 2]
```diff
@@ -349,43 +380,23 @@ std::map<std::string, std::function<void()>> construct_tests() {
...
+   auto model = ContainerList()
+     .append(Linear(2, 8).make())
+     .append(Linear(8, 1).make())
+     .make();
 
-     running_loss = running_loss * 0.99 + loss.data().sum().toCFloat() * 0.01;
-     EXPECT(epoch < 3000);
-     epoch++;
-   }
+   auto optim = Adagrad(model, 1.0).make();
```
Comment (ebetica) [inline test.cpp]: I would add lr_decay and weight_decay in here as well.

File: autograd.cpp (hunk 1 above)
Comment (eugene-kharitonov) [reply autograd.cpp]: Ok, just copied from SGD.

File: test.cpp (hunk 2 above)
Comment (eugene-kharitonov) [reply test.cpp]: Fixed

**Pull request merged**